### PR TITLE
[WIP] Support RR360 (Rotated Rectangle 360) Detection in TRR360D

### PR DESCRIPTION
**Pull request by vansin**
## 1. Motivation

传统的 mmrotate的 oc le90 le135 定义的算法和数据集检测范围不支持 360的旋转模板检测，不能区分矩形的ABCD点的语义，所以本工作，定义了 r360 角度定义法，同时制作了DOTA格式的TRR360D 数据集，在此数据集上验证改造算法的有效性。

image

## 2. Modification

- A new generated dataset （TRR360D ）
- R360 angle define, q2rbox r2qbox funcition
- RR360LocalVisualizer
- DOTAR360Metric

## 3. 快速上手

```shell
mkdir RR360
cd RR360
export WORK_HOME=$(pwd)
```

安装前置依赖
```shell
conda create -n RR360 python=3.8 -y
conda activate RR360

pip install torch==1.10.0+cu111 torchvision==0.11.0+cu111 torchaudio==0.10.0 -f https://download.pytorch.org/whl/torch_stable.html

pip install -U openmim
mim install mmengine
mim install "mmcv>=2.0.0rc1"

cd $WORK_HOME
git clone https://github.com/open-mmlab/mmdetection -b dev-3.x
cd mmdetection 
pip install -v -e .

```

```shell
# 假设已经安装mmengine、mmcv 2.x、mmdetection
cd $WORK_HOME
git clone https://github.com/vansin/mmrotate -b RR360
cd mmrotate 
export MMROTATE_HOME=$(pwd)
pip install -v -e .
```

数据集下载

```shell
cd $MMROTATE_HOME
mkdir -p data/TRR360D
# git clone https://github.com/vansin/TRR360D.git data/TRR360D
wget https://openmmlab.vansin.top/datasets/ICDAR2019_MTD_HOQ.zip -O data/TRR360D/TRR360D.zip
cd $MMROTATE_HOME/data/TRR360D
unzip TRR360D.zip
```

600 个训练图片，240个测试图片

通过 browse_dataset.py 检测数据集的正确性

```shell
cd $MMROTATE_HOME
 python projects/RR360/tools/browse_dataset.py \
 projects/RR360/configs360/rotated_retinanet/rotated-retinanet-rbox-r360_r50_fpn_6x_ic19.py \
    --output-dir show_ann \
    # --show
```

数据集有由语义信息的，ABCD点是表格左上角开始，依次顺时钟的四个点。
image

# configs360

推理已训练好的模型

```shell
cd $MMROTATE_HOME

wget https://openmmlab.vansin.top/work_dirs/RR360/configs360/rotated_rtmdet_x3_r/rotated_rtmdet_s_l1-3x-ic19_pt/epoch_36.pth

python projects/RR360/tools/test.py \
    projects/RR360/configs360/rotated_rtmdet_x3_r/rotated_rtmdet_s_l1-3x-ic19_pt.py \
    epoch_36.pth \
    --show-dir predict_result
    # --show

```
在 work_dirs/rotated_rtmdet_s_l1-3x-ic19_pt/{时间戳}/predict_result 可以看到训练好的模型推理结果
image

训练

```shell
cd $MMROTATE_HOME
python projects/RR360/tools/train.py \
    projects/RR360/configs360/rotated_rtmdet_x3_r/rotated_rtmdet_s_l1-3x-ic19_pt.py \
    --d
```

测试自行训练的

```shell
cd $MMROTATE_HOME
python projects/RR360/tools/test.py \
    work_dirs/RR360/configs360/rotated_rtmdet_x3_r/rotated_rtmdet_s_l1-3x-ic19_pt/rotated_rtmdet_s_l1-3x-ic19_pt.py \
    work_dirs/RR360/configs360/rotated_rtmdet_x3_r/rotated_rtmdet_s_l1-3x-ic19_pt/epoch_36.pth
```

### CODE REVIEW

File: mmrotate/models/task_modules/coders/delta_xywht_hbbox_coder.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def __init__(self,
         self.means = target_means
         self.stds = target_stds
         self.angle_version = angle_version
-        assert self.angle_version in ['oc', 'le135', 'le90']
+        assert self.angle_version in ['oc', 'le135', 'le90', 'h180']
```
Comment (zytx121) [inline mmrotate/models/task_modules/coders/delta_xywht_hbbox_coder.py:57]: `360` may be more appropriate.

File: projects/RR360/configs360/_base_/default_runtime.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+# default_scope = 'mmrotate'
+
+# default_hooks = dict(
```
Comment (zytx121) [inline projects/RR360/configs360/_base_/default_runtime.py:3]: Delete unnecessary comments.

File: projects/RR360/hooks/fps_hook.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
+import numpy as np
+
+from mmengine.registry import HOOKS
+from mmengine.hooks import Hook
+from mmengine.device import get_max_cuda_memory, is_cuda_available
+
+import pymongo
```
Comment (zytx121) [inline projects/RR360/hooks/fps_hook.py:7]: Tell users  in README.md that he needs to install `pymongo`.

File: projects/RR360/models/losses/a_l1_loss.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+import numpy as np
+from mmrotate.registry import MODELS
+from mmdet.models.losses.utils import weighted_loss
+
+@weighted_loss
+def al1_loss(pred: Tensor, target: Tensor) -> Tensor:
```
Comment (zytx121) [inline projects/RR360/models/losses/a_l1_loss.py:12]: Tell users the difference between the two losses in README.md

**Comment by zytx121** [comment]
`configs360` is empty now.

File: mmrotate/models/task_modules/coders/delta_xywht_hbbox_coder.py (hunk 1 above)
Comment (vansin) [reply mmrotate/models/task_modules/coders/delta_xywht_hbbox_coder.py:57]: h360 ?

Comment (zytx121) [reply mmrotate/models/task_modules/coders/delta_xywht_hbbox_coder.py:57]: OK

Comment (vansin) [reply mmrotate/models/task_modules/coders/delta_xywht_hbbox_coder.py:57]: r360 ？

File: projects/RR360/hooks/fps_hook.py (hunk 3 above)
Comment (vansin) [reply projects/RR360/hooks/fps_hook.py:7]: delete the benchmark so that won't use pymongo

**Comment by makramhamad** [comment]
@vansin & @zytx121 any chances this of merging PR? I kind of stuck in a similar issue too.

**Comment by vansin** [comment]
> @vansin & @zytx121 any chances of merging this PR? I kind of stuck in a similar issue. Another thing is this implemented for all models and dataset?

Only can be used for Rotated Rectangle Datasets, and must be label with DOTA Similar with limitations ABCD points [x1, y1, x2, y2, x3, y3, x4,y4, label, Difficult]

**Comment by makramhamad** [comment]
> > @vansin & @zytx121 any chances of merging this PR? I kind of stuck in a similar issue. Another thing is this implemented for all models and dataset?
> 
> Only can be used for Rotated Rectangle Datasets, and must be label with DOTA Similar with limitations ABCD points [x1, y1, x2, y2, x3, y3, x4,y4, label, Difficult]

That is good, there no restriction on model right? I want to use if for redet model. When will it be live any ideas?

**Comment by vansin** [comment]
> > > @vansin & @zytx121 any chances of merging this PR? I kind of stuck in a similar issue. Another thing is this implemented for all models and dataset?
> > 
> > 
> > Only can be used for Rotated Rectangle Datasets, and must be label with DOTA Similar with limitations ABCD points [x1, y1, x2, y2, x3, y3, x4,y4, label, Difficult]
> 
> That is good, there no restriction on model right? I want to use if for redet model. When will it be live any ideas?

I just try on rtmdet and retinanet , redet may need to change some params

**Comment by vansin** [comment]
> > > > @vansin & @zytx121 any chances of merging this PR? I kind of stuck in a similar issue. Another thing is this implemented for all models and dataset?
> > > 
> > > 
> > > Only can be used for Rotated Rectangle Datasets, and must be label with DOTA Similar with limitations ABCD points [x1, y1, x2, y2, x3, y3, x4,y4, label, Difficult]
> > 
> > 
> > That is good, there no restriction on model right? I want to use if for redet model. When will it be live any ideas?
> 
> I just try on rtmdet and retinanet , redet may need to change some params

may before 4 month, Now there is a higher priority

**Comment by makramhamad** [comment]
> 

Thanks for quick response and help. Looking forward to use this.

File: projects/RR360/models/losses/a_l1_loss.py (hunk 4 above)
Comment (vansin) [reply projects/RR360/models/losses/a_l1_loss.py]: I Delete the AL1 loss , I find L1 loss achieve high mAP

**Review by zytx121 — APPROVED** [review]
LGTM

**Comment by zytx121** [comment]
@RangiLyu hope merge as soon as possible

File: projects/RR360/readme.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
+# RR360(360 Rotated Rectangle Detection)
+
+本项目在TRR360D数据集上支持360度旋转矩形框目标检测
```
Comment (RangiLyu) [inline projects/RR360/readme.md:3]: Please also provide an English version readme.

Comment (vansin) [reply projects/RR360/readme.md]: done

File: projects/RR360/tools/test.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+from mmengine.config import Config, DictAction
+from mmengine.evaluator import DumpResults
+from mmengine.registry import RUNNERS
+from mmengine.runner import Runner
+from projects.RR360.structures.bbox import RotatedBoxes
+
+import mmrotate.structures
+from mmrotate.utils import register_all_modules
+
+mmrotate.structures.bbox.RotatedBoxes = RotatedBoxes
```
Comment (RangiLyu) [inline projects/RR360/tools/test.py:16]: 这种替换package里面模块的方式不太好，是不是所有要用这个模型的地方都得在脚本前面加上这句话？有其他替代的办法吗？如果暂时没有的话就先留个TODO吧。

Comment (vansin) [reply projects/RR360/tools/test.py:16]: 暂时没有哟，好的留个 TODO，主要是融入到已有的代码，动的地方比较多， RotatedBoxes很多地方都是直接实例话的，不是用config build 出来的，导致不能修改 config实现，后面有时间把  RotatedBoxes 弄成 config build 方式吧

**Comment by vansin** [comment]
@RangiLyu 加TODO 的意思是要readme 写一下吗？

**Comment by RangiLyu** [comment]
> @RangiLyu 加TODO 的意思是要readme 写一下吗？

代码里加上 `TODO:xxxxx` 的注释说明一下这么做是暂时的，之后会修改

**Comment by RangiLyu** [comment]
> @RangiLyu 加TODO 的意思是要readme 写一下吗？

Readme最好也说明一下情况

**Comment by vansin** [comment]
> > @RangiLyu 加TODO 的意思是要readme 写一下吗？
> 
> Readme最好也说明一下情况

ok

**Review by RangiLyu — APPROVED** [review]

**Comment by lzh420202** [comment]
Does this coding directly apply to existing models to improve or degrade performance? I am also concerned about how to accurately evaluate this method, since the actual orientation information is not effectively represented in the mAP.

**Comment by vansin** [comment]
> Does this coding directly apply to existing models to improve or degrade performance? I am also concerned about how to accurately evaluate this method, since the actual orientation information is not effectively represented in the mAP.

please see~

https://github.com/open-mmlab/mmrotate/blob/65a9a2c4a40027e05ec7f7b2043d699726d541ea/projects/RR360/evaluation/functional/mean_ap.py#L95

https://github.com/open-mmlab/mmrotate/blob/dev-1.x/projects/RR360/evaluation/metrics/dota_r360_metric.py

**Comment by lzh420202** [comment]
> > Does this coding directly apply to existing models to improve or degrade performance? I am also concerned about how to accurately evaluate this method, since the actual orientation information is not effectively represented in the mAP.
> 
> please see~
> 
> https://github.com/open-mmlab/mmrotate/blob/65a9a2c4a40027e05ec7f7b2043d699726d541ea/projects/RR360/evaluation/functional/mean_ap.py#L95
> 
> https://github.com/open-mmlab/mmrotate/blob/dev-1.x/projects/RR360/evaluation/metrics/dota_r360_metric.py
> 
> dota_r360_metric

Thanks, I understand the evaluation method, but did you test it on DOTA1.0? I just want to know if it has any advantages over the le90, le135, and oc on RS.
Thanks for your attention.

**Comment by vansin** [comment]
> > > Does this coding directly apply to existing models to improve or degrade performance? I am also concerned about how to accurately evaluate this method, since the actual orientation information is not effectively represented in the mAP.
> > 
> > 
> > please see~
> > https://github.com/open-mmlab/mmrotate/blob/65a9a2c4a40027e05ec7f7b2043d699726d541ea/projects/RR360/evaluation/functional/mean_ap.py#L95
> > 
> > https://github.com/open-mmlab/mmrotate/blob/dev-1.x/projects/RR360/evaluation/metrics/dota_r360_metric.py
> > dota_r360_metric
> 
> Thanks, I understand the evaluation method, but did you test it on DOTA1.0? I just want to know if it has any advantages over the le90, le135, and oc on RS. Thanks for your attention.

not test in DOTA 1.0, define a  new  metric named r360
